core/runtime/src/main/java/io/quarkus/runtime/CleanableExecutor.java

```diff
@@ -0,0 +1,254 @@
+package io.quarkus.runtime;
+
+import java.lang.reflect.Field;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.jboss.threads.EnhancedQueueExecutor;
+
+import sun.misc.Unsafe;
+
+/**
+ * An executor that can be 'cleaned'. Each clean increments an internal generation number. When a task is executed by
+ * a thread that has a different generation number to the last task executed by the thread then the threads ThreadLocal's
+ * will be cleared. This means that thread locals from a previous deployment cannot interfere with the current deployment.
+ *
+ * This is only for development mode, it must not be used for production applications.
+ *
+ * TODO: should this just provide a facacde that simply starts a new thread pool instead?
```

Comment:
Maybe that would be a bot cleaner: just start a new pool, swap it in, and shut down the old one.

Reply:
you would need to also re-dispatch the current HTTP thread to the new pool, to prevent old context from hanging around.

Reply:
This approach works for now, this approach was listed as a todo because it would need a lot more testing (which is mostly manual testing at the moment as we do not have a good way to test dev mode). 

This is a very low risk approach as it is basically the status quo with some fixes. Changing the behaviour of dev mode and hot reload has much more potential for regressions, which is why I left it as a TODO item.